lib/puppet/type/file.rb

```diff
@@ -656,25 +677,18 @@ def recurse_remote(children)
       found = []
       total.reject! do |data|
         result = found.include?(data.relative_path)
-        found << data.relative_path unless found.include?(data.relative_path)
+        found << data.relative_path unless result
```

Comment:
I'd say the overall behavior is unchanged, but I also try not to make optimizations while refactoring, so I'd lean towards keeping it out of the first commit.